### ISSUE TITLE
Could not find a schema at lastCachedGraphQLSchema.json

### ISSUE THREAD

**Issue by nitinlitmus7** [magento/pwa-studio#1052]
I run the "yarn run build" command and  I faced that given below issue. Please help me out. Thanks:-

```
:/var/www/html/magento23/pwa-studio$ yarn run build
yarn run v1.15.2
$ yarn workspaces run build
$ echo 'Skipping graphql-cli-validate-magento-pwa-queries build...'
Skipping graphql-cli-validate-magento-pwa-queries build...
$ concurrently --raw yarn:build:cjs yarn:build:esm
$ BABEL_ENV=development babel src --out-dir esm --root-mode 'upward' --source-maps
$ BABEL_ENV=production babel src --out-dir dist --root-mode 'upward' --source-maps
Successfully compiled 32 files with Babel.
Successfully compiled 49 files with Babel.
$ yarn run clean
$ rimraf dist
$ babel src --out-dir dist --root-mode 'root' --source-maps --copy-files
Successfully compiled 16 files with Babel.
$ yarn run clean && yarn run build:esm && yarn run validate-queries && yarn run build:prod
$ rimraf dist esm
$ BABEL_ENV=development babel src --out-dir esm --root-mode 'upward' --source-maps --copy-files
Successfully compiled 290 files with Babel.
$ yarn run download-schema && graphql validate-magento-pwa-queries --project venia
$ graphql get-schema --project venia
⚠ maximum redirect reached at: http://magento.local2/admin_rx200y/admin/index/index/key/d82ea5208d0cc660efbf78d01e078c0004844e19a1ad9079f8d58196170a9eea/
Validating GraphQL queries in venia project...
✖ An error occurred:
Could not find a schema at lastCachedGraphQLSchema.json.
 Run 'graphql get-schema --project venia' to download the schema before running validate-magento-pwa-queries.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed.
Exit code: 1
Command: /usr/bin/node
Arguments: /usr/share/yarn/lib/cli.js build
Directory: /var/www/html/magento23/pwa-studio/packages/venia-concept
Output:

info Visit https://yarnpkg.com/en/docs/cli/workspaces for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.`
```

**Screenshots**
![Screenshot from 2019-03-22 21-54-52](https://user-images.githubusercontent.com/46247890/54837848-5f385c00-4ced-11e9-833b-399585d17469.png)
![Screenshot from 2019-03-22 21-55-56](https://user-images.githubusercontent.com/46247890/54837849-5f385c00-4ced-11e9-9d16-36cef943ad6b.png)

**Please complete the following device information:**
- Ubuntu - 18.04
 - Magento Version - 2.3
-  Yarn v 1.15.2
 - NPM version `npm -v` - 6.7
 - Node Version `node -v` - 11.12

**Comment by sirugh** [comment]
@nitinlitmus7 what is your `MAGENTO_BACKEND_URL`? The most recent release, and `develop` branch, will only work on Magento `2.3.1`.

**Comment by nitinlitmus7** [comment]
@sirugh  thanks for reply, i installed the magento2.3 in ubuntu system locally and i gave my MAGENTO_BACKEND_URL is 'http://magento.local2/admin_rx200y' and please explain which branch i need to develop to run this, Thanks

**Comment by sirugh** [comment]
@nitinlitmus7 The latest `develop` or `release/2.1` branches will only work with Magento `2.3.1`. If your Magento backend is `2.3.0` you will need to run the `2.0` tag of this project. Install the `v2.0.0` version and try to run it again.

`git checkout v2.0.0` should do the trick.

Let me know if that works.

**Comment by nitinlitmus7** [comment]
@sirugh thanks for reply, i figured out that issue on morning and it is running fine.
Thanks

**Comment by nitinlitmus7** [comment]
@sirugh  if i have run `yarn run watch:venia` that command working properly but if i run `yarn run watch:all` that command will give error. Can you please help me out of this :
`Using environment variables from .env
Error: [pwa-buildpack:Utilities:configureHost.js] Could not setup development domain: 
Error: Creating a local development domain requires an interactive terminal for the user to answer prompts. Run the development server (e.g. `yarn run watch:venia`) by itself in the terminal to continue.
    at Promise (/var/www/html/magento23/pwa-studio-2.0.0/packages/pwa-buildpack/dist/Utilities/configureHost.js:67:25)
    at new Promise (<anonymous>)
    at getCert (/var/www/html/magento23/pwa-studio-2.0.0/packages/pwa-buildpack/dist/Utilities/configureHost.js:55:10)
    at configureHost (/var/www/html/magento23/pwa-studio-2.0.0/packages/pwa-buildpack/dist/Utilities/configureHost.js:162:18)
    at Object.configure (/var/www/html/magento23/pwa-studio-2.0.0/packages/pwa-buildpack/dist/WebpackTools/PWADevServer.js:139:17)
    at configureHost (/var/www/html/magento23/pwa-studio-2.0.0/packages/pwa-buildpack/dist/Utilities/configureHost.js:165:11)`

**Comment by sirugh** [comment]
@nitinlitmus7 if you have other issues please open a new issue. I think other folks would really appreciate if you could post _how_ you fixed your issue.

**Comment by sirugh** [comment]
Oh I see, you're saying that you were using the `admin` url for your `MAGENTO_BACKEND_URL` instead of your site url. Cool. Closing this issue.

**Comment by nitinlitmus7** [comment]
@sirugh ok, 
I installed Magento 2.3.0 with PWA Studio 2.1 and PWA studio 2.1 is not compatible with Magento 2.3.0, so i downloaded PWA Studio 2.0. Run `yarn run build` command and it is working fine. The main issue was PWA and Magento incompatible.
Thanks

**Comment by supernova-at** [comment]
Just chiming in for posterity here, but as part of `yarn build` PWA Studio issues a GraphQL [introspection query](https://graphql.org/learn/introspection/) to the endpoint specified in `MAGENTO_BACKEND_URL` and saves the resulting schema in `lastCachedGraphQLSchema.json` (it then validates all of the project's queries against that schema).

So the steps are:
1. Download the schema and save it as `lastCachedGraphQLSchema.json`
2. Validate the queries against that schema

A lot of the discussion so far on this issue are around PWA / Magento compatibility, and those issues certainly do come up as part of Step 2 - that's the explicit intended purpose of the `validate-queries` build step.

But the issue described here looks more like the process failed on Step 1. The `MAGENTO_BACKEND_URL` in `.env` is definitely the place to start, and I'm glad that solved your problem 👍

**Comment by supernova-at** [comment]
I should also note that we don't have much control over Step 1 but we do have [an open PR](https://github.com/graphql-cli/graphql-cli/pull/447) into `graphql-cli` to fix up their error reporting when something goes wrong.

Once the PR is merged on their end we can output some better error messages when this issue occurs.

**Comment by nitinlitmus7** [comment]
@supernova-at thanks for reply, issue was resolved, it was a PWA Studio version issue.
Thanks

**Comment by siddhrock** [comment]
> `MAGENTO_BACKEND_URL` - `magento admin url` to `magento site url`.

@nitinlitmus7 , have you done any change apart from this ? I did same but no luck.  Please do needful.

**Comment by nitinlitmus7** [comment]
@siddhrock exactly what error you are facing. Please give the error detail. Thanks

**Comment by siddhrock** [comment]
I am using Magento2.3.1 with PWA 2.1.0 

error which I am facing is 


$ yarn run build                                                                                                                                                                                yarn run v1.15.2                                                                                                                                                                                $ yarn workspaces run build                                                                                                                                                                     $ echo 'Skipping graphql-cli-validate-magento-pwa-queries build...'                                                                                                                             'Skipping graphql-cli-validate-magento-pwa-queries build...'                                                                                                                                    $ concurrently --raw yarn:build:cjs yarn:build:esm                                                                                                                                              $ BABEL_ENV=development babel src --out-dir esm --root-mode 'upward' --source-maps                                                                                                              $ BABEL_ENV=production babel src --out-dir dist --root-mode 'upward' --source-maps                                                                                                              'BABEL_ENV' is not recognized as an internal or external command,                                                                                                                               operable program or batch file.                                                                                                                                                                 error Command failed with exit code 1.                                                                                                                                                          info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.                                                                                                            'BABEL_ENV' is not recognized as an internal or external command,                                                                                                                               operable program or batch file.                                                                                                                                                                 error Command failed with exit code 1.                                                                                                                                                          info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.                                                                                                            error Command failed with exit code 1.                                                                                                                                                          info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.                                                                                                            error Command failed.                                                                                                                                                                           Exit code: 1                                                                                                                                                                                    Command: C:\Program Files\nodejs\node.exe                                                                                                                                                       Arguments: C:\Program Files (x86)\Yarn\lib\cli.js build                                                                                                                                         Directory: C:\wamp64\www\pwa-studio\packages\peregrine                                                                                                                                          Output:                                                                                                                                                                                                                                                                                                                                                                                         info Visit https://yarnpkg.com/en/docs/cli/workspaces for documentation about this command.                                                                                                     error Command failed with exit code 1.                                                                                                                                                          info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
--

**Comment by siddhrock** [comment]
@nitinlitmus7 
I am using Magento2.3.1 with PWA 2.1.0

Error which I am facing is  
![image](https://user-images.githubusercontent.com/48045578/55875912-e291f280-5bb3-11e9-828b-1391dd9cf82a.png)

**Comment by nitinlitmus7** [comment]
@siddhrock which platform you are using, I am using Ubuntu-18.04. If you are using windows it has some issues- "https://magento.stackexchange.com/questions/264730/pwa-setup-in-magento-2-3-0-on-windows-7". If you will try on ubuntu that will not give error.

**Comment by siddhrock** [comment]
@nitinlitmus7  I  am using Windows 10 Pro

**Comment by patrickpeng2091** [comment]
hello, 
yarn run build error , how to fix ?
Magento 2 CE 2.3.1 ,
nodejs:v12.2.0
yarn:1.16.0
my MAGENTO_BACKEND_URL : http://magento2.s1.com/

 request to http://magento2.s1.com//graphql failed, reason: getaddrinfo ENOTFOUND magento2.s1.com
Validating GraphQL queries in venia project...
✖ An error occurred:
Could not find a schema at lastCachedGraphQLSchema.json.
 Run `graphql get-schema --project venia` to download the schema before running validate-magento-pwa-queries.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed.
Exit code: 1
Command: /usr/bin/node
Arguments: /usr/share/yarn/lib/cli.js build
Directory: /var/www/magento2/app/design/frontend/pwa-studio/packages/venia-concept
Output:

info Visit https://yarnpkg.com/en/docs/cli/workspaces for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

**Comment by DanielRuf** [comment]
Bump, still the same.

Is it due to self-signed certs in my case?

```
⚠ request to https://magento.localhost/graphql failed, reason: unable to verify the first certificate
Validating GraphQL queries in venia project...
✖ An error occurred:
Could not find a schema at lastCachedGraphQLSchema.json.
```

**Comment by DanielRuf** [comment]
Hi @patrickpeng2091,

please check if you can access the domain.

> getaddrinfo ENOTFOUND magento2.s1.com

**Comment by dodiya-divyarajsinh** [comment]
Hi All, I face the same issue. Please see screenshot.
https://i.stack.imgur.com/ZsGio.png

![ZsGio](https://user-images.githubusercontent.com/25686742/58853352-98ab2180-86b7-11e9-9efb-b2d46961f67c.png)

**Comment by sirugh** [comment]
@dodiya-divyarajsinh can you execute `which graphql` in your terminal and show me the output? What you want is to use the graphql installed by `pwa-studio` located within your node_modules:

```
❯ which graphql
./node_modules/.bin/graphql
```

**Comment by dodiya-divyarajsinh** [comment]
Hi, @sirugh thank for the response,
It will show nothing(blank) as follow

![image](https://user-images.githubusercontent.com/25686742/58930825-d6bd4980-877a-11e9-8648-b83a41579d20.png)

**Comment by DanielRuf** [comment]
And `stat node_modules/.bin/graphql`?

**Comment by dodiya-divyarajsinh** [comment]
![image](https://user-images.githubusercontent.com/25686742/58940537-50633080-8797-11e9-9852-cdf3af1437ab.png)

I have use YARN

**Comment by LucasCalazans** [comment]
Same issue here.
Magento -> 2.3.1
PWA-Studio -> develop

After running `yarn run build`

```
yarn run v1.16.0
$ yarn workspaces run build
$ cp ../../magento-compatibility.js ./lib/
$ concurrently --raw yarn:build:cjs yarn:build:esm
$ BABEL_ENV=production babel src --out-dir dist --root-mode 'upward' --source-maps
$ BABEL_ENV=development babel src --out-dir esm --root-mode 'upward' --source-maps
Successfully compiled 44 files with Babel.
Successfully compiled 70 files with Babel.
$ yarn run clean
$ rimraf dist
$ babel src --out-dir dist --root-mode 'root' --source-maps --copy-files
Successfully compiled 12 files with Babel.
$ yarn run clean && yarn run build:esm && yarn run validate-queries && yarn run build:prod
$ rimraf dist esm
$ BABEL_ENV=development babel src --out-dir esm --root-mode 'upward' --source-maps --copy-files
Successfully compiled 312 files with Babel.
$ yarn run download-schema && graphql validate-magento-pwa-queries --project venia
$ graphql get-schema --project venia
⚠ request to https://magento-pwa.localhost/graphql failed, reason: certificate has expired
Validating GraphQL queries in venia project...
✖ An error occurred:
Could not find a schema at lastCachedGraphQLSchema.json.
 Run `graphql get-schema --project venia` to download the schema before running validate-magento-pwa-queries.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed.
Exit code: 1
Command: /home/me/.nvm/versions/node/v10.15.1/bin/node
Arguments: /usr/share/yarn/lib/cli.js build
Directory: /home/me/projects/magento-pwa/venia/packages/venia-concept
Output:

info Visit https://yarnpkg.com/en/docs/cli/workspaces for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by tjwiebell** [comment]
@LucasCalazans - It appears we're not allowing insecure backends in query validation, which is why you're seeing this error. I think it's reasonable to add the `insecure: true` flag to `packages/venia-concept/.graphqlconfig` so this validation works with self-signed certificates, especially since we only expect to validate queries in development environments (skipped in production builds). I'll update the issue description with this proposed fix.

In the meantime, if you change your `MAGENTO_BACKEND_URL` to http, this should get around the error, or you can manually add this flag to `.graphqlconfig`.

**Comment by LucasCalazans** [comment]
I think that the error I'm getting is the same reported:

```
✖ An error occurred:
Could not find a schema at lastCachedGraphQLSchema.json.
```

**Comment by threescs** [comment]
> 嗨@DanielRuf，
> 下面是 `stat node_modules/.bin/graphql`
> 
> ![图片](https://user-images.githubusercontent.com/25686742/58940537-50633080-8797-11e9-9852-cdf3af1437ab.png)
> 
> 我用YARN



> @ nitinlitmus7 我使用的是Windows 10 Pro
hello,I am the same problem as you, have you solved it?

**Comment by threescs** [comment]
> 你好，
> 纱线运行构建错误，如何解决？
> Magento 2 CE 2.3.1，
> nodejs：v12.2.0 
> yarn：1.16.0 
> my MAGENTO_BACKEND_URL：[http](http://magento2.s1.com/)：//magento2.s1.com/[](http://magento2.s1.com/)
> 
> 请求http://magento2.s1.com//graphql失败，原因：getaddrinfo ENOTFOUND magento2.s1.com 
> 在venia项目中验证GraphQL查询... 
> ✖发生错误：
> 无法在lastCachedGraphQLSchema.json中找到架构。
> 运行`graphql get-schema --project venia`运行验证，Magento的-PWA-查询之前下载的模式。
> 错误命令失败，退出代码为1. 
> info 有关此命令的文档，请访问https://yarnpkg.com/en/docs/cli/run。
> 错误命令失败，退出代码为1. 
> info 有关此命令的文档，请访问https://yarnpkg.com/en/docs/cli/run。
> 错误命令失败。
> 退出代码：1 
> 命令：/ usr / bin / node
> 参数：/usr/share/yarn/lib/cli.js build 
> 目录：/ var / www / magento2 / app / design / frontend / pwa-studio / packages / venia-concept 
> 输出：
> 
> info 有关此命令的文档，请访问https://yarnpkg.com/en/docs/cli/workspaces。
> 错误命令失败，退出代码为1. 
> info 有关此命令的文档，请访问https://yarnpkg.com/en/docs/cli/run。

I am the same problem as you, have you solved it?

**Comment by mrRonyTran** [comment]
Hey guys,

I also got the same errors.

Ubuntu 18.04
Magento 2.3.2 CE
yarn 1.16.0
node 10.16.0
npm 6.9.0
@magento/pwa-studio 3.0.0


$ graphql get-schema --project venia --insecure
⚠ request to https://magentoce232pwa.local/graphql failed, reason: getaddrinfo ENOTFOUND magentoce232pwa.local magentoce232pwa.local:443
Validating GraphQL queries in venia project...
✖ An error occurred:
Could not find a schema at lastCachedGraphQLSchema.json.

**Comment by awilcoxa** [comment]
@zetlen to weigh in

**Comment by zetlen** [comment]
If this is still happening to you, would you mind posting the output of `curl https://magentoce232pwa.local/graphql`?

**Comment by awilcoxa** [comment]
Closing this bug assuming the issue no longer occurs, if it does, please re-open the bug and provide repro steps.
